M2/Macaulay2/packages/KGAlgebras.m2

```diff
@@ -0,0 +1,191 @@
+            productInv := position(matrixElts,M -> M==matrixElts#i * matrixElts#j);
+            H := (FA_i)*(FA_j)-(FA_productInv);
+            multTableIdealGens#((n+1)*i+j) = H;
+        );
+    ); 
+    FA/ideal(toList multTableIdealGens) -- output
+);
+random(FreeAlgebraQuotient) := opt -> (S) -> (       
+    randelt := sum(apply(toList(0 .. #(gens S)-1), i -> random(coefficientRing S) * S_i))
+);
+--True implies irreducible. Otherwise returns proper K[G]-submodule. 
+meataxe = method()
+meataxe(FiniteGroupAction, FreeAlgebraQuotient,Ring) := (G,S,K) -> (
+    while true do(
+        --Do the meataxe algorithm 
+        p := random S; --generate random element of KG algebra
+        L := toList coefficients p;  
+        M := (group G)#0-(group G)#0; -- zero matrix which has the same dimensions as an element of G
+        m := 0;
+        for x in (flatten entries L#0) do(
+            n := index(x);
+            M = M + (group G)#n * sub((flatten entries L#1)#m,K);
+            m = m + 1    
+        ); --This for loop generates a matrix representing p using the matrix elements for G
+        --Factors of the characteristic polynomial of M
+        q := apply(factor characteristicPolynomial M, y-> value y); --Get an array containing the factors of the characteristic polynomial of the matrix M
+        q = sort toList q; --Sorts q by degree
+        if (degree q#0)#0 == 0 then(
+            q = drop(q,1)--Drops constant factor if possible
+        );
+        
+        Mp := (group G)#0-(group G)#0; --zero matrix same dimensions as elements of G
+        
+        for P in q do( 
+            n := (degree P)#0; 
+            MP := Mp = (group G)#0-(group G)#0; 
+            
+            --Evaluates P on the matrix M
+            for m in 0 .. #apply(apply((entries (coefficients P)#0)#0,degree),x-> x#0)-1 do(
+                Mp=M^((apply(apply((entries (coefficients P)#0)#0,degree),x-> x#0))#m);
+                Mp=Mp*(apply((entries (coefficients P)#1),x-> x#0))#m;
+                MP=MP+Mp;
+            );
+            
+
+            --MP is p(theta) in the algorithm
+            N := kernel MP;
+            y := (gens N)_0;--Takes a "random" element of the kernel
+            M1 := matrix apply(group G, z-> z*y); --multiplies all group elements by the random element, giving column vectors, and we make a matrix M1 from those vectors
+            if rank M1 != rank ((group G)#0) then (
+                return image M1;
+            );--Checks that the module generated by M1 is not a strict submodule of KG
+            
+            H := apply(group G, transpose); --Generate list of group elements' matrices but transposed
+            MPP := transpose MP;
+            NP := kernel MPP;
+            yp := (gens NP)_0;
+            M2 := matrix apply(H,z -> z*yp);
+            if debugLevel == 1 then (
+                print(rank M2);
+            );
+            if rank M2 != rank (H#0) then (
+                return image M2;
```

Comment:
Sometimes this function returns a boolean and sometimes it returns a module.  That seems confusing for the caller -- how do they know what type to expect?

Would it make sense to return both as a sequence?  So the caller could do something like:

```m2
(isIrreducible, M) = meataxe(G,S,K)
if isIrreducible then ...
```

Reply:
I have another implementation of the meataxe for modules and we return a list of the indecomposable pieces. So if the input M is indecomposable, the ouput is just {M}. You could try this instead.